File: src/types/codec.zig
const types = @import("main.zig");
const std = @import("std");
/// An interface for protocol-specific data encoding.
///
/// Invariants:
///
/// - `encode()` and `decode()` are inverse functions; that is, `decode(encode(x)) == x`.
pub fn Codec(comptime I: type) type {
    return union(enum) {
        varInt: types.VarInt(I),
        pub fn decode(self: Codec(I), reader: std.io.reader(u8)) !I {
            return switch (self) {
                inline else => |s| s.decode(reader),
            };
        }

        pub fn encode(self: Codec(I), writer: anytype) !void {
            switch (self) {
                inline else => |s| s.encode(writer),
            }
        }
    };
}

File: src/types/Codec.rs
use std::{
    io::{Read, Write},
    num::NonZeroUsize,
};

/// An interface for the data types defined by the Minecraft server protocol.
pub(crate) trait Codec<T> {
    const MAX_SIZE: NonZeroUsize;
    fn encode(&self, writer: &mut impl Write) -> Result<(), ()>;
    fn decode(reader: &mut impl Read) -> Result<T, ()>;
}
